docs/rfcs/rfc03.md

```diff
@@ -0,0 +1,127 @@
+
+## Background
+
+Jsonnet goes way out of its way to ensure that all the imports are well known statically, and that the evaluation is fully reproducible
+and cacheable by providing side-effect free evaluation and avoiding computed imports.
+
+Kubecfg adds the ability to directly import external dependencies. This is meant to be used for importing immutable external files,
+usually links to some resource at some specific git sha. It also [supports](https://github.com/kubecfg/kubecfg/pull/218) explicitly enforcing a fingerprint check.
+
+However, fetching external dependencies comes with problems:
+
+* downloding external dependencies may be slow and annoying
+* external files may disappear permanently or temporarily
+
+## Proposal
+
+We propose adding a download cache that will store a local copy of files downloaded from remote sources.
+
+The cache directory structure, is suitable for checking-in the cache into the current VCS repo (if the user so chooses).
+
+### Cache structure
+
+Let's consider this jsonnet source file:
+
+```jsonnet
+import 'https://github.com/kubecfg/k8s-libsonnet/raw/144e62ccbdc9f7d4f8bccc14f33224cbf63e5185/k8s.libsonnet'
+```
+
+The cache directory can be set via the `--cache-dir` flag. By default its `<cachebase>/kubecfg`, where `<cachebase>` is determined by the [os.UserCacheDir](https://pkg.go.dev/os@go1.20.2#UserCacheDir).
+
+
+After the first execution, kubecfg will create this directory structure
+
+```console
+$ tree ~/Library/Caches/kubecfg
+/Users/mkm/Library/Caches/kubecfg
+└── https
+    └── github.com
+        └── kubecfg
+            └── k8s-libsonnet
+                └── raw
+                    └── 144e62ccbdc9f7d4f8bccc14f33224cbf63e5185
+                        └── k8s.libsonnet
+
+
+### Multiple cache sources
+
+(optional, advanced)
+
+The user can define multiple cache directories.
+They will act as a "search path".
+If a file is not found in the first directory, the next one will be tried until the cache directories are exhausted.
+Only the first cache directory is used to write new files to.
+
+Example:
+
+```bash
+kubecfg --cache-dir ./vendor --cache-dir ~/.cache/kubecfg show ....
+```
+
+Open questions:
+
+* should we always put the default cache at the end of the cache search path?
+
+### Vendoring
+
+To enable vendoring the user needs to pass a custom location for the cache directory and choose a desirable cache pruning policy.
+
+This plays well with the [Flags From Files](https://github.com/kubecfg/kubecfg/pull/224) feature.
+
+```console
+$ cat .kubecfgrc.jsonnet
+{
+    flags+: {
+        'cache-dir': 'vendor',
+        'cache-pruning-policy': 'vendor'
+    },
+}
+$ kubecfg show foo.jsonnet
+```
+
+### Cache pruning
+
+The simplest thing to do is to let the cache just grow and grow and let the user just clean it up when needed.
+
+The other strategies are:
+
+* size limit, delete least recently used
+* remove all unreferenced
+* ???
+
+#### Remove all unreferenced strategy
```

Comment:
Thanks for writing this up.  This seems way harder than a "remove older than X" subcommand, and as you point out it's impossible unless we can enumerate _all_ the jsonnet top-levels using the shared cache somehow.

Reply:
Hrm. Another version is change the above "write to first" to instead _always_ write to first.  That way we can 'rotate' the cache dirs by adding a new cache dir first, that reads from the existing cache dirs; run all kubecfg we care about with the same new cache dir; then eventually replace the old cache dir with the content of the new cache dir (or flip flags to achieve the same thing).

This suggests that `--cache-dir ./vendor` always updates `./vendor` with a _copy_ of files used by this repo, pulled from any existing homedir-wide cache if there's a cache hit.   I think that's what people expect from `./vendor`, although this also suggests each repo pulls down its own copy without ever storing a copy in home-dir cache - which could be undesirable if I had many such repos.  I guess a power user can create a hierarchy of `--cache-dir ./vendor --cache-dir ../other/vendor`, if they care about this.  Another option is to _always_ write to the homedir-wide cache (unless disabled) in addition to the first `--cache-dir`.

(Fwiw, my prior imaginary simplest-thing-that-could-work version of this whole feature just had a single `--cache-dir` that defaulted to `~/.cache/kubecfg`.  If you change the value to `./vendor` then you're also implicitly disabling your homedir cache.  No cache eviction, we just encourage `rm -r; rerun` if folks care.  Real simple, real dumb.  I still kind of like this option because it's so easy to describe.)